### PR TITLE
Stack arrays along new axis

### LINKED ISSUE #28: dask.array.concatenate
Really two issues here

1.  We should implement a `concatenate` function that concatenates many `dask.Array` objects into a single one.  This is relatively straightforward.
2.  We need to think about how users should specify collections of arrays as a single `dask.Array` 

For example given a directory of netCDF files, each with several datasets, we might want something magical like the following

    into(da.Array, '/data/myfile.*.h5::/data/path/to/X', blockshape=(1, 1000, 1000))

Or maybe this is too magical.  Maybe we need to have a more verbose function that allows users to specify their intent more precisely.  One approach with `concatenate` would be

    uris = [fn + '::/data/path/to/X' for fn in sorted(glob('./data/myfile.*.h5'))]
    arrays = [into(Array, uri, blockshape=(1000, 1000)) for uri in uris]
    arr = da.stack(arrays, axis=0)

I'd value @shoyer 's opinion here.

### PR DESCRIPTION
**Pull request by mrocklin**
Example
-------

Create slices

```Python
>>> import dask.array as da
>>> import numpy as np

>>> data = [da.into(da.Array, np.ones((4, 4)), blockshape=(2, 2))
...          for i in range(3)]

>>> x = stack(data, axis=0)
>>> x.shape
(3, 4, 4)

>>> stack(data, axis=1).shape
(4, 3, 4)

>>> stack(data, axis=2).shape
(4, 4, 3)
```

cc @shoyer

Fixes #28

### CODE REVIEW

**Comment by mrocklin** [comment]
I believe that this can be used repeatedly with nested lists with no loss of performance.

File: dask/array/tests/test_array_core.py [hunk 1]
```diff
@@ -154,3 +154,24 @@ def test_keys():
...
+    assert s.blockdims == ((1, 1, 1), (2, 2), (3, 3))
+    assert s.dask[(s.name, 0, 1, 0)] == ('A', 1, 0)
+    assert s.dask[(s.name, 2, 1, 0)] == ('C', 1, 0)
+
+    s2 = stack([a, b, c], axis=1)
+    assert s2.shape == (4, 3, 6)
+    assert s2.blockdims == ((2, 2), (1, 1, 1), (3, 3))
+    assert s2.dask[(s2.name, 0, 1, 0)] == ('B', 0, 0)
+    assert s2.dask[(s2.name, 1, 1, 0)] == ('B', 1, 0)
+
```
Comment (shoyer) [inline dask/array/tests/test_array_core.py]: It would be nice to see tests for stacking along `axis=2` and `axis=-1`

Comment (mrocklin) [inline dask/array/tests/test_array_core.py]: Good idea.  Will do.

**Comment by mrocklin** [comment]
Also added `concatenate`

File: dask/array/tests/test_array_core.py [hunk 2]
```diff
@@ -154,3 +154,46 @@ def test_keys():
...
+    assert x.dask[(x.name, 0, 1)] == ('A', 0, 1)
+    assert x.dask[(x.name, 5, 0)] == ('C', 1, 0)
+
+    y = concatenate([a, b, c], axis=1)
+
+    assert y.shape == (4, 18)
+    assert y.blockdims == ((2, 2), (3, 3, 3, 3, 3, 3))
+    assert y.dask[(y.name, 1, 0)] == ('A', 1, 0)
+    assert y.dask[(y.name, 1, 5)] == ('C', 1, 1)
+
```
Comment (shoyer) [inline dask/array/tests/test_array_core.py]: same comment about testing for negative `axis` here.

also, error handling for an invalid `axis`?

File: dask/array/core.py [hunk 3]
```diff
@@ -736,3 +737,119 @@ def get(dsk, keys, get=threaded.get, **kwargs):
...
+    >>> da.stack(data, axis=1).shape
+    (4, 3, 4)
+
+    Result is a new dask Array
+
+    See Also:
+        concatenate
+    """
+    n = len(seq)
+    assert len(set(a.blockdims for a in seq)) == 1  # same blockshape
```
Comment (shoyer) [inline dask/array/core.py]: Is there some sort of a utility for aligning blockshapes?

**Comment by mrocklin** [comment]
> Is there some sort of a utility for aligning blockshapes?

We could do this.  Perhaps shape is to `reshape` as blockshape is to *reblock*?  I'll probably punt on this until it's clear people are using `dask.array` though.  I want to make sure I'm not prematurely developing this.

Until then we'll need to be consistent at creation time.

File: dask/array/tests/test_array_core.py (hunk 1 above)
Comment (mrocklin) [inline dask/array/tests/test_array_core.py]: Done.

File: dask/array/tests/test_array_core.py (hunk 2 above)
Comment (mrocklin) [inline dask/array/tests/test_array_core.py]: Done.

**Pull request merged**